src/python/pants/engine/rules.py

```diff
@@ -413,21 +422,35 @@ class Rule(ABC):
     def output_type(self):
         """An output `type` for the rule."""
 
-    @property
-    @abstractmethod
-    def dependency_rules(self):
-        """A tuple of @rules that are known to be necessary to run this rule.
 
-        Note that installing @rules as flat lists is generally preferable, as Rules already
-        implicitly form a loosely coupled RuleGraph: this facility exists only to assist with
-        boilerplate removal.
-        """
+def register_rules(*namespaces: Union[ModuleType, Mapping[str, Any]]) -> Iterable[Rule]:
+    """Registers all @rules in the given namespaces.
 
-    @property
-    @abstractmethod
-    def dependency_optionables(self):
-        """A tuple of Optionable classes that are known to be necessary to run this rule."""
-        return ()
+    If no namespaces are given, registers all the @rules in the caller's module namespace.
+    """
+    if not namespaces:
+        currentframe = inspect.currentframe()
+        assert isinstance(currentframe, FrameType)
+        caller_frame = currentframe.f_back
+        caller_module = inspect.getmodule(caller_frame)
+        assert isinstance(caller_module, ModuleType)
+        namespaces = (caller_module.__dict__,)
```

Comment:
Note to self - do not call `.__dict__` here; instead fix a bug by iterating `namespaces` after the if check and converting any ModuleType namespaces to Mappings via `.__dict__`.

Reply:
Fixed